### PR TITLE
Ctrl/Cmd+S Saves Code to localStorage

### PR DESCRIPTION
**Pull request by jakethecoder95**
So I did try to do as little changes as possible to the current code.  I created a new action creator of type; "saveCode", and a new code-storage-epic; "userManualSaveEpic".  I also refactored "saveCodeEpic" to "saveOnExecuteChallengeEpic".  The event listener is added and removed in the classic/Editor component.

Before this, the users code was getting saved to localStorage when they attempted to submit the challenge. I've just changed it so that they can manually do it as well.  I kept pressing ctrl+s accidently so I thought it would be cool to add this feature. Please let me know if this can be improved in any way.  I am brand new to open source so I apologize if my approach has not been perfect. If you guys like this idea, It would also be cool to add a small alert message to show the user their codes been saved locally. Just thought I'd start with this and see if people like the idea.

### CODE REVIEW

**Comment by thecodingaviator** [comment]
Please remove the lockfile from the PR

**Comment by jakethecoder95** [comment]
Got it. Thanks

**Comment by raisedadead** [comment]
Thanks for the pull request. Can you point me to any issue thread or forum post if this was discussed for more context.

**Comment by raisedadead** [comment]
Ignore previous comment. Found it in your explanation in OP:

> Before this, the users code was getting saved to localStorage when they attempted to submit the challenge. I've just changed it so that they can manually do it as well. I kept pressing ctrl+s accidently so I thought it would be cool to add this feature. [...]

**Comment by jakethecoder95** [comment]
> Ignore previous comment. Found it in your explanation in OP:
> 
> > Before this, the users code was getting saved to localStorage when they attempted to submit the challenge. I've just changed it so that they can manually do it as well. I kept pressing ctrl+s accidently so I thought it would be cool to add this feature. [...]

File: client/src/templates/Challenges/redux/code-storage-epic.js [hunk 1]
```diff
@@ -71,17 +78,26 @@ function clearCodeEpic(action$, state$) {
...
+      return !isCodeLockedSelector(state$.value);
+    }),
+    tap(() => saveToLocalStorage(state$)),
+    ignoreElements()
+  );
+}
+
+function saveOnExecuteChallengeEpic(action$, state$) {
   return action$.pipe(
     ofType(types.executeChallenge),
```
Comment (Bouncey) [inline client/src/templates/Challenges/redux/code-storage-epic.js]: `ofType` accepts multiple action types so `ofType(types.executeChallenge, types.saveCode)` will run the epic on execute and save, reducing the duplication

File: client/src/templates/Challenges/redux/index.js [hunk 2]
```diff
@@ -142,6 +143,7 @@ export const unlockCode = createAction(types.unlockCode);
 export const disableBuildOnError = createAction(types.disableBuildOnError);
 export const storedCodeFound = createAction(types.storedCodeFound);
 export const noStoredCodeFound = createAction(types.noStoredCodeFound);
+export const saveCode = createAction(types.saveCode);
```
Comment (Bouncey) [inline client/src/templates/Challenges/redux/index.js]: I think `saveEditorContents` is more descriptive. `code` can mean many things given a different context.

File: client/src/templates/Challenges/redux/index.js [hunk 3]
```diff
@@ -239,6 +242,10 @@ export const reducer = handleActions(
       ...state,
       challengeFiles: payload
     }),
+    [types.saveCode]: (state, { payload }) => ({
```
Comment (Bouncey) [inline client/src/templates/Challenges/redux/index.js]: Why update the store? Does the file change?

File: client/src/templates/Challenges/redux/index.js [hunk 4]
```diff
@@ -221,6 +223,7 @@ export const reducer = handleActions(
       ...state,
       challengeIdToNameMap: payload
     }),
+
```
Comment (Bouncey) [inline client/src/templates/Challenges/redux/index.js]: Extra whitespace

**Review by Bouncey — CHANGES_REQUESTED** [review]

File: client/src/templates/Challenges/redux/index.js (hunk 3 above)
Comment (jakethecoder95) [reply client/src/templates/Challenges/redux/index.js]: Yeah this was not necessary.

**Comment by raisedadead** [comment]
@jakethecoder95 thanks a lot for the PR. 

We also need to warn the users that something from the localStorage will be loaded on the editor. This can be a notification.

Also, can you cleanup the PR by squashing your commits. (I know we can do it while merging, but it just makes QA easy, when I pull this down).

**Comment by jakethecoder95** [comment]
@raisedadead Hey, appreciate the feedback.  

What kind of notification do you mean? Like a small alert saying "Saved to localStorage", similar to what codepen does?

**Comment by raisedadead** [comment]
> What kind of notification do you mean? Like a small alert saying "Saved to localStorage", similar to what codepen does?

Yes, a toast would be nice.

**Comment by raisedadead** [comment]
Hi @jakethecoder95 I am going to have to close this PR. Looks like the rebase got really weird.

We would love you to still open a fresh PR if you have the commits on a branch safe somewhere?

**Pull request closed without merging**